File: ambeth/Ambeth.Merge/ambeth/util/setup/AbstractDatasetBuilder.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using De.Osthus.Ambeth.Log;
using De.Osthus.Ambeth.Ioc.Annotation;
using De.Osthus.Ambeth.Merge;

#if SILVERLIGHT
using De.Osthus.Ambeth.Util;
#else
using System.Threading;
#endif


namespace De.Osthus.Ambeth.Util.Setup
{
    public abstract class AbstractDatasetBuilder : IDatasetBuilder
    {
        private static readonly ThreadLocal<IList<Object>> INITIAL_TEST_DATASET_TL = new ThreadLocal<IList<Object>>();

        [LogInstance]
        public ILogger Log { private get; set; }

	    [Autowired]
	    protected IEntityFactory EntityFactory;

	    [Autowired]
	    protected IEntityMetaDataProvider EntityMetaDataProvider;

        protected abstract void BuildDatasetInternal();
        
        
        public IList<IDatasetBuilder> GetDependsOn()
        {
            return null;
        }

        public void BuildDataset(IList<Object> initialTestDataset)
        {
            IList<Object> oldSet = INITIAL_TEST_DATASET_TL.Value;
            INITIAL_TEST_DATASET_TL.Value = initialTestDataset;
            try
            {
                BuildDatasetInternal();
            }
            finally
            {
                if (oldSet != null)
                {
                    INITIAL_TEST_DATASET_TL.Value = oldSet;
                }
                else
                {
                    INITIAL_TEST_DATASET_TL.Value = null;
                }
            }
        }


        protected T CreateEntity<T>()
	    {
		    T entity = EntityFactory.CreateEntity<T>();
            INITIAL_TEST_DATASET_TL.Value.Add(entity);
		    return entity;
	    }
    }
}

File: jambeth/jambeth-merge/src/main/java/de/osthus/ambeth/util/setup/AbstractDatasetBuilder.java
package de.osthus.ambeth.util.setup;

import java.util.Collection;

import de.osthus.ambeth.collections.IdentityHashSet;
import de.osthus.ambeth.ioc.annotation.Autowired;
import de.osthus.ambeth.log.ILogger;
import de.osthus.ambeth.log.LogInstance;
import de.osthus.ambeth.merge.IEntityFactory;
import de.osthus.ambeth.merge.IEntityMetaDataProvider;
import de.osthus.ambeth.merge.IMergeProcess;

public abstract class AbstractDatasetBuilder implements IDatasetBuilder
{
	@LogInstance
	private ILogger log;

	@Autowired
	protected IEntityFactory entityFactory;

	@Autowired
	protected IEntityMetaDataProvider entityMetaDataProvider;

	@Autowired
	protected IMergeProcess mergeProcess;

	protected final ThreadLocal<Collection<Object>> initialTestDatasetTL = new ThreadLocal<Collection<Object>>();

	@Override
	public Collection<Object> buildDataset()
	{
		beforeBuildDataset();
		try
		{
			buildDatasetInternal();
			return initialTestDatasetTL.get();
		}
		finally
		{
			afterBuildDataset();
		}
	}

	protected abstract void buildDatasetInternal();

	protected void beforeBuildDataset()
	{
		IdentityHashSet<Object> initialTestDataset = new IdentityHashSet<Object>();
		initialTestDatasetTL.set(initialTestDataset);
	}

	protected void afterBuildDataset()
	{
		initialTestDatasetTL.remove();
	}

	protected <V> V createEntity(Class<V> entityType)
	{
		V entity = entityFactory.createEntity(entityType);
		initialTestDatasetTL.get().add(entity);
		return entity;
	}
}
